### PR TITLE
Fix and refactor label models

### PR DESCRIPTION
**Pull request by fayejf**
# What does this PR do ?

Fix and refactor label models

**Collection**: ASR

# Changelog 
- fix hanging issue happens in test after train in` speech_to_label.py `, `speaker_reco.py` and `speaker_reco_finetune`.py
- fix weighted cross entropy by disabling cal_label_occurences for val and test
- refactor loss instantiation part in label_models 
- updated yamls file for above refactory.

# Usage
Specify loss as below in config files for label_models
```python
loss:
    _target_: nemo.collections.common.losses.cross_entropy.CrossEntropyLoss
    weight: 'auto' # could be 'auto' or 1D tensor or null
# or
loss:
    _target_: nemo.collections.asr.losses.angularloss.AngularSoftmaxLoss # you could also use cross-entrophy loss
    scale: 30
    margin: 0.2
    
```
# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/asr/conf/lang_id/titanet_large.yaml [hunk 1]
```diff
@@ -145,12 +144,10 @@ model:
     num_classes: 107
     pool_mode: 'xvector' # 'attention'
     emb_sizes: 512
-    angular: False
 
   loss:
-    weight: 'auto' # could either be 'auto' or 1D tensor
-    scale: 30 # specify this if use angular loss
-    margin: 0.2  # specify this if use angular loss
```
Comment (nithinraok) [inline examples/asr/conf/lang_id/titanet_large.yaml:153]: can you also update gradient_clip_val to 1.0

File: examples/speaker_tasks/recognition/conf/SpeakerNet_recognition_3x2x512.yaml [hunk 2]
```diff
@@ -101,9 +101,9 @@ model:
     num_classes: 7205
     pool_mode: 'xvector'
     emb_sizes: [512,512]
-    angular: False
 
   loss:
+    _target_: from nemo.collections.common.losses.CrossEntropyLoss # you could also use AngularSoftmaxLoss
```
Comment (nithinraok) [inline examples/speaker_tasks/recognition/conf/SpeakerNet_recognition_3x2x512.yaml:106]: remove scale and margin for cross entrophy loss.

File: examples/asr/conf/lang_id/titanet_large.yaml (hunk 1 above)
Comment (fayejf) [reply examples/asr/conf/lang_id/titanet_large.yaml:153]: Sure.

File: nemo/collections/asr/models/label_models.py [hunk 3]
```diff
@@ -87,7 +87,18 @@ def list_available_models(cls) -> List[PretrainedModelInfo]:
 
     def __init__(self, cfg: DictConfig, trainer: Trainer = None):
         self.world_size = 1
-        self.cal_labels_occurrence = False
+        self.cal_labels_occurrence_train = False
+        self.labels_occurrence = None
+
+        if 'loss' in cfg:
```
Comment (nithinraok) [inline nemo/collections/asr/models/label_models.py:93]: You may assume loss in cfg but check if 'weight'. in params and if its there check for auto. Also check if pretrained models working or not

File: nemo/collections/asr/models/label_models.py [hunk 4]
```diff
@@ -97,29 +108,15 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
-                    else:
-                        weight = cfg.loss.weight
-                    self.loss = CELoss(weight=weight)
-                else:
-                    self.loss = CELoss()
+        if self.labels_occurrence:
+            # Goal is to give more weight to the classes with less samples so as to match the ones with the higher frequencies
+            weight = [sum(self.labels_occurrence) / (len(self.labels_occurrence) * i) for i in self.labels_occurrence]
+
+        if 'loss' not in cfg:
```
Comment (nithinraok) [inline nemo/collections/asr/models/label_models.py:115]: why is this necessary again? put in else case above

File: nemo/collections/asr/models/label_models.py [hunk 5]
```diff
@@ -321,7 +339,8 @@ def training_step(self, batch, batch_idx):
     def evaluation_step(self, batch, batch_idx, dataloader_idx: int = 0, tag: str = 'val'):
         audio_signal, audio_signal_len, labels, _ = batch
         logits, _ = self.forward(input_signal=audio_signal, input_signal_length=audio_signal_len)
-        loss_value = self.loss(logits=logits, labels=labels)
+        # evaluation step (validation, test) loss is CE loss
```
Comment (nithinraok) [inline nemo/collections/asr/models/label_models.py:342]: why this comment?

File: nemo/collections/asr/models/label_models.py (hunk 4 above)
Comment (fayejf) [reply nemo/collections/asr/models/label_models.py:115]: it serves to instantiate loss when there is no loss in config file, such as unit test

Comment (fayejf) [reply nemo/collections/asr/models/label_models.py:115]: yeah you are right, we can put it as 
```
if self.labels_occurrence:
    weight = [sum(self.labels_occurrence) / (len(self.labels_occurrence) * i) for i in self.labels_occurrence]
else:
    self.loss = CrossEntropyLoss()  # default loss for this class, basically to pass test
```

Comment (fayejf) [reply nemo/collections/asr/models/label_models.py:115]: you mean here? 

```
if 'loss' in cfg:
    xxx
else:
    self.loss = CrossEntropyLoss()  # default loss for this class, basically to pass test
    self.eval_loss = CrossEntropyLoss()
```

we cannot assign module before Module.__init__() call

File: nemo/collections/asr/models/label_models.py (hunk 5 above)
Comment (fayejf) [reply nemo/collections/asr/models/label_models.py:342]: oops...I forgot to remove it

File: nemo/collections/asr/models/label_models.py [hunk 6]
```diff
@@ -87,39 +86,62 @@ def list_available_models(cls) -> List[PretrainedModelInfo]:
...
         if trainer is not None:
             self.world_size = trainer.num_nodes * trainer.num_devices
 
         super().__init__(cfg=cfg, trainer=trainer)
 
-        self.preprocessor = EncDecSpeakerLabelModel.from_config_dict(cfg.preprocessor)
-        self.encoder = EncDecSpeakerLabelModel.from_config_dict(cfg.encoder)
-        self.decoder = EncDecSpeakerLabelModel.from_config_dict(cfg.decoder)
+        if self.labels_occurrence:
+            # Goal is to give more weight to the classes with less samples so as to match the ones with the higher frequencies
```
Comment (nithinraok) [inline nemo/collections/asr/models/label_models.py:107]: do we need this comment?

File: nemo/collections/asr/models/label_models.py [hunk 7]
```diff
@@ -87,39 +86,62 @@ def list_available_models(cls) -> List[PretrainedModelInfo]:
...
-                    if cfg.loss.weight == 'auto':
-                        self.cal_labels_occurrence = True
-                        # Goal is to give more weight to the classes with less samples so as to match the ones with the higher frequencies
-                        weight = [
-                            sum(self.labels_occurrence) / (len(self.labels_occurrence) * i)
-                            for i in self.labels_occurrence
-                        ]
+            # To support older version checkpoints
+            if '_target_' not in cfg.loss:
+                logging.info("Support older version checkpoints of label models")
```
Comment (nithinraok) [inline nemo/collections/asr/models/label_models.py:113]: instead of this logging message, can you add:
setting angular=False in decoder is deprecated and will be removed in 1.13 version, use specific loss with `_target_`

**Review by nithinraok — APPROVED** [review]
small changes, rest LGTM.

File: nemo/collections/asr/models/label_models.py (hunk 7 above)
Comment (fayejf) [reply nemo/collections/asr/models/label_models.py:113]: Sure

File: nemo/collections/asr/models/label_models.py (hunk 6 above)
Comment (fayejf) [reply nemo/collections/asr/models/label_models.py:107]: let's keep it. This is to let user to understand how we calculate weight easily

**Review by nithinraok — APPROVED** [review]
